\newglossaryentry{transportome}{%
  name={transportome},
  description={All the membrane transporters and channels that govern influx and efflux of ions in a cell}
}

\newglossaryentry{batch effect}{%
  name={batch effect},
  description={A difficult to define term that describes the overall technical bias introduced in an experiment which is visible only when comparing samples treated in different batches, or in different experiments}
}

\newglossaryentry{library}{%
  name={library},
  description={In the context of the wet-lab: A gene library is a collection of (labelled) RNA molecules or DNA fragments from a specific source, such as a tissue or a cell line. In the context of computing: A code library (or package) is an external collection of functions that can be installed and used to provide particular functionalities},
  plural={libraries}
}

\newglossaryentry{docstring}{%
  name={docstring},
  description={A comment that describes the input, output and functionality of a piece of code, most often a single function, but also a module (a collection of functions).}
}

\newglossaryentry{computational reproducibility}{%
  name={computational reproducibility},
  description={For a bioinformatic analysis to be computationally reproducible, it must produce, given the same inputs, the same outputs. Many factors can give computationally irreproducible results, such as differences in the hardware, kernel, \gls{OS}, or environment variables in each computer. For instance, dependency management is one of the main issues to overcome to achieve C.R}
}

\newglossaryentry{exon cluster}{%
  name={exon cluster},
  description={A group of one or more probe sets that cover a contiguous stretch of putatively transcribed genomic sequence, typically corresponding to an exon within a transcript (from the \href{https://www.affymetrix.com/support/help/exon_glossary/index.affx}{Affimetrix glossary page})}
}

\newglossaryentry{transcript cluster}{%
  name={transcript cluster},
  description={A group of one or more exon clusters (on Exon Arrays) or probes (on Gene Arrays) covering a region of the genome reflecting all the exonic transcription evidence known for the region and corresponding to a known or putative gene. The underlying exonic evidence can come from transcripts of well-annotated genes or predicted genes (from the \href{https://www.affymetrix.com/support/help/exon_glossary/index.affx}{Affimetrix glossary page})}
}

\newglossaryentry{regex}{%
  name={RegEx},
  description={Short for Regular Expression. A sequence of characters that define a pattern that can be interpreted by search engines to find regions of text in strings. For example, the pattern "\mono{\^{}ba.*}" matches any string that begins with the characters \mono{ba}}
}


\newacronym{deg}{DEG}{Differentially Expressed Gene}
\newacronym[
    longplural={Differential Expression Analyses}
]{dea}{DEA}{Differential Expression Analysis}
\newacronym{mdp}{MDP}{Microarray Data Preprocessing}
\newacronym{gsea}{GSEA}{Gene Set Enrichment Analysis}
\newacronym{bart}{BART}{Bioinformatics Array Research Tool}
\newacronym{gattaca}{GATTACA}{General Algorithm for The Transcriptional Analysis by one-Channel Arrays}
\newacronym{os}{OS}{Operating System}
\newacronym{pca}{PCA}{Principal Component Analysis}
\newacronym{wsl}{WSL}{Windows Subsystems for Linux}
\newacronym{yaml}{YAML}{YAML Ain't Markup Language}
\newacronym{rma}{RMA}{Robust Multichip Average}
\newacronym{fdr}{FDR}{False Discovery Rate}
\newacronym{pfp}{PFP}{Proportion of False Positives}
\newacronym{gam}{GAM}{Generalized Additive Model}
\newacronym{hugo}{HUGO}{\underline{Hu}man \underline{G}enome \underline{O}rganization}
\newacronym{geo}{GEO}{Gene Expression Omnibus}
